File: scripts/validate-prompt.js
#!/usr/bin/env node
// validate-prompt.js
// Simple Node.js script to validate the minimal prompt JSON schema used by the
// documentation prompt library. It's intended as a developer convenience and
// is not part of Goose runtime.
// Usage: node scripts/validate-prompt.js path/to/prompt.json
const fs = require('fs');
const path = require('path');

function validatePrompt(filePath) {
  const raw = fs.readFileSync(filePath, 'utf8');
  let data;
  try {
    data = JSON.parse(raw);
  } catch (err) {
    console.error(`JSON parse error for ${filePath}:`, err.message);
    process.exitCode = 2;
    return false;
  }

  const required = ['id','title','description','category','example_prompt','example_result','extensions'];
  const missing = required.filter(k => !(k in data));
  if (missing.length) {
    console.error(`Missing required fields in ${filePath}:`, missing.join(', '));
    process.exitCode = 3;
    return false;
  }

  if (!Array.isArray(data.extensions) || data.extensions.length < 3) {
    console.error(`'extensions' must be an array with 3 or more items (found ${Array.isArray(data.extensions)?data.extensions.length:'not an array'})`);
    process.exitCode = 4;
    return false;
  }

  for (let i=0;i<data.extensions.length;i++){
    const ext = data.extensions[i];
    if (!ext.name || !ext.description || typeof ext.is_builtin === 'undefined' || !Array.isArray(ext.environmentVariables)){
      console.error(`Extension at index ${i} is missing required fields (name, description, is_builtin, environmentVariables)`);
      process.exitCode = 5;
      return false;
    }
  }

  // check title matches filename-ish
  const fname = path.basename(filePath, '.json');
  const normalizedTitle = data.title.toLowerCase().replace(/[^a-z0-9]+/g,'-').replace(/(^-|-$)/g,'');
  if (normalizedTitle !== fname) {
    console.warn(`Warning: normalized title '${normalizedTitle}' does not match filename '${fname}'. This is recommended but not required.`);
  }

  console.log(`VALID: ${filePath} — ${data.title}`);
  return true;
}

const args = process.argv.slice(2);
if (!args.length) {
  console.error('Usage: validate-prompt.js <path-to-prompt.json>');
  process.exitCode = 1;
  process.exit(1);
}

const ok = validatePrompt(args[0]);
process.exit(ok?0:1);

File: scripts/validate-prompt.py
#!/usr/bin/env python3
"""
scripts/validate-prompt.py
Simple Python validator for prompt JSON files in the prompt library.
This is a convenience script to check required fields and extension counts
before opening a PR. It is not executed by Goose itself.

Usage:
    python3 scripts/validate-prompt.py path/to/prompt.json
"""
import sys
import json
from pathlib import Path

def validate(file_path):
    p = Path(file_path)
    if not p.exists():
        print(f"File not found: {file_path}")
        return 2
    try:
        data = json.loads(p.read_text())
    except Exception as e:
        print(f"JSON parse error for {file_path}: {e}")
        return 3

    required = ['id','title','description','category','example_prompt','example_result','extensions']
    missing = [k for k in required if k not in data]
    if missing:
        print(f"Missing required fields: {', '.join(missing)}")
        return 4

    exts = data.get('extensions')
    if not isinstance(exts, list) or len(exts) < 3:
        print(f"'extensions' must be a list with 3+ items (found {type(exts).__name__} with length {len(exts) if isinstance(exts,list) else 'N/A'})")
        return 5

    for i,ext in enumerate(exts):
        if not all(k in ext for k in ('name','description','is_builtin','environmentVariables')):
            print(f"Extension at index {i} missing required fields (name, description, is_builtin, environmentVariables)")
            return 6
        if not isinstance(ext.get('environmentVariables'), list):
            print(f"Extension environmentVariables must be a list at index {i}")
            return 7

    # small filename check
    fname = p.stem
    normalized_title = ''.join(c.lower() if c.isalnum() else '-' for c in data.get('title','')).strip('-')
    if normalized_title != fname:
        print(f"Warning: normalized title '{normalized_title}' does not match filename '{fname}' — recommended but optional.")

    print(f"VALID: {file_path} — {data.get('title')}")
    return 0

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: validate-prompt.py <file.json>')
        sys.exit(1)
    rc = validate(sys.argv[1])
    sys.exit(rc)
